sanitycheck: Use logger and avoid using global options

nashif opened the pull request:
Move away from custom output functions and use logging module instead. Cleanup
most usage of global options and instead pass arguments and/or use class
variables. This is needed to make unittesting of sanitycheck easier, right now
everything is being controlled via command line options.

nashif committed:
sanitycheck: move hardware map generation out of main

Move all hardware map generation/usage to a seperate class. This will
make it easier to extend the supported hardware in the future.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: handle flashing issues with --device-testing

When something goes bad during the flashing process set the reason
correctly and put the error messages from the flasher into device.log
and display the location of that file instead of an empty handler.log
right now.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
requirements: require tabulate and anytree modules for sanitycheck

tabulate: Needed to show list of devices in a well formatted table.
anytree: needed to list testsuite in a tree form.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: output hardware registery after scanning

Show a list of all refistered devices after scanning with

sanitycheck --generate-hardware-map

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: fix typo

retyring -> retrying
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
saniytcheck: fixed pylint issues

Fixed a few pylint issues, including space and indent reported by
flake8.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: move job setting to main()

Report about number of jobs in the main() function and not in the class.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: use outdir from TS instance

Stop using global options and instead use propagated variable.
This is to make testing of sanitycheck easier.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: cleanup setting binary options

Stop using global options and instead set binary options using class
variables.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: use logging module for all console output

Drop custom output functions and use logging mode for almost all
reporting. Also log everything into sanitycheck.log so detailed run
information can be inspected later, even if we did not run with
--verbose mode on the console.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: reduce usage of global options

The use of a global options variable is making it very difficult to
create a testsuite for this script, so reduce and push options into
arguments instead.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: reenable --timestamps option with loggging module

--timestamps now just enables a different formatter with asctime.
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: use logger instead of print

Use logger instead of print for consistency.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: consistent output from logger

Make all output consistent with logger and report errors correctly.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: move global info() into Testsuite class

This function is not used anywhere else now that we use logging module,
so push this into the class where it is being used.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: all output from flasher goes to logger.debug

Redirect all output from flasher to logger.debug().

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: do not use gloabl options

More changes moving away from using global options and instead using
class variables and parameters.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: move log_info_* to ProjectBuilder class

Move this to where it is being used and called.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: get rid of more global options

More global option cleanup.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: remove unused globals

A few globals that were never used anymore.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: pylint issues

run through pylint and fix.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

maksimmasalski commented:
@nashif will you add --pre-test and --post-test options to the sanitycheck to let before and after west flash run additional program?

nashif commented:
> @nashif will you add --pre-test and --post-test options to the sanitycheck to let before and after west flash run additional program?

after this PR get merged, and I am not adding --pre-test and --post-test, those will have to be per platform and part of the hardware map for that platform.

maksimmasalski commented:
> > @nashif will you add --pre-test and --post-test options to the sanitycheck to let before and after west flash run additional program?
> 
> after this PR get merged, and I am not adding --pre-test and --post-test, those will have to be per platform and part of the hardware map for that platform.

Do I need to start modifying sanitycheck script to add that kind of options to let Microchip board be added into the Sanitycheck testing? Microchip requires that options. Because during programming board must be in the reset mode. If nobody will add that kind of options, Microchip will not work with Sanitycheck.

maksimmasalski commented on scripts/sanitycheck:
```diff
@@ -3710,34 +3737,71 @@ def native_and_unit_first(a, b):
-run_individual_tests = None
-options = None
+class HardwareMap:
+    manufacturer = [
```
For the Microchip board necessary to add a support, because I found out that I can't create a hardware map for that board, it is not supported. That board uses runner `dediprog` for west flash, but even in current Sanitycheck script it is not added.
So necessary to add one more option `Microchip`

maksimmasalski commented on scripts/sanitycheck:
```diff
@@ -3710,34 +3737,71 @@ def native_and_unit_first(a, b):
+        'NXP Semiconductors'
+    ]
+
+    runner_mapping = {
```
In runner_mapping necessary to add
`dediprog`:  [ 'Dediprog SF100' ]

maksimmasalski commented:
> > > @nashif will you add --pre-test and --post-test options to the sanitycheck to let before and after west flash run additional program?
> > 
> > 
> > after this PR get merged, and I am not adding --pre-test and --post-test, those will have to be per platform and part of the hardware map for that platform.
> 
> Do I need to start modifying sanitycheck script to add that kind of options to let Microchip board be added into the Sanitycheck testing? Microchip requires that options. Because during programming board must be in the reset mode. If nobody will add that kind of options, Microchip will not work with Sanitycheck.

Now I understand what do you mean about hardware map. Generate hardware map for the Microchip board, and then in the hardware map write path to the program, then sanitycheck can read hardware map and execute that program. Sorry for my misunderstanding. Unfortunately, hardware map is not available for the Microchip board.

maksimmasalski commented:
@nashif  
I have a performance issue.
1. During running of your branch's sanitycheck
`sanitycheck --device-testing --device-serial /dev/ttyACM0 -p reel_board -T ../tests/kernel/fifo -v -v`
it freezes in that place 
`DEBUG   - run test: reel_board/tests/kernel/fifo/fifo_api/kernel.fifo.poll
DEBUG   - run test: reel_board/tests/kernel/fifo/fifo_api/kernel.fifo
DEBUG   - run test: reel_board/tests/kernel/fifo/fifo_usage/kernel.fifo.usage
DEBUG   - run test: reel_board/tests/kernel/fifo/fifo_timeout/kernel.fifo.timeout
DEBUG   - run test: reel_board/tests/kernel/fifo/fifo_timeout/kernel.fifo.timeout.poll
DEBUG   - run test: reel_board/tests/kernel/fifo/fifo_usage/kernel.fifo.usage.poll
`
2. Run of the sanitycheck
`sanitycheck --device-testing --device-serial /dev/ttyACM0 -p reel_board -T ../tests/kernel/critical/ -v -v`
freezes too `DEBUG   - Building /home/maksim/zephyrproject/zephyr/tests/kernel/critical for reel_board
DEBUG   - run test: reel_board/tests/kernel/critical/kernel.common.critical`

Build log says:
` [ 96%] Linking C executable zephyr_prebuilt.elf
 Memory region         Used Size  Region Size  %age Used
            FLASH:       32732 B         1 MB      3.12%
             SRAM:        7448 B       256 KB      2.84%
         IDT_LIST:          72 B         2 KB      3.52%
 [ 96%] Built target zephyr_prebuilt
 Scanning dependencies of target linker_pass_final_script_target
 [ 96%] Generating linker_pass_final.cmd                                          
 [ 96%] Built target linker_pass_final_script_target
 [ 97%] Generating isr_tables.c
 Scanning dependencies of target zephyr_final
 [ 98%] Building C object zephyr/CMakeFiles/zephyr_final.dir/misc/empty_file.c.obj
 [ 98%] Building C object zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj
 [100%] Linking C executable zephyr.elf
 Generating files from zephyr.elf for board: reel_board
 [100%] Built target zephyr_final`

**What can be the reason that during the run process just got stuck?**

JordanYates commented:
While reworking the logging functionality, would it also be possible to fix the issue with colors on Windows with cmd.exe?
https://github.com/zephyrproject-rtos/zephyr/issues/21409

nashif commented:
> Do I need to start modifying sanitycheck script to add that kind of options to let Microchip board be added into the Sanitycheck testing? Microchip requires that options. Because during programming board must be in the reset mode. If nobody will add that kind of options, Microchip will not work with Sanitycheck.

I am not sure how this is related to this PR.

nashif committed:
sanitycheck: fix legacy device testing support

When testing with one device and without a hardware map, make sure we
load the temporary map into the suite.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@maksimmasalski  try again.

nashif commented:
> While reworking the logging functionality, would it also be possible to fix the issue with colors on Windows with cmd.exe?

yeah, will take a look at this after this gets merged.

nashif committed:
sanitycheck: use colorama module for colors

Use colorama module for colors instead of custom codes.

Fixes #21409

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> yeah, will take a look at this after this gets merged.

just fixed this here :)

maksimmasalski commented:
> @maksimmasalski try again.

Now it works.

andrewboie commented on scripts/requirements.txt:
```diff
@@ -1,5 +1,6 @@
 Pillow
 PyYAML>=5.1
+anytree
```
oh good, I was about to send a PR to add this :)

andrewboie commented on scripts/sanitycheck:
```diff
@@ -229,24 +228,19 @@ sys.path.insert(0, os.path.join(ZEPHYR_BASE, "scripts/"))
-    COLOR_YELLOW = ""
+
+logger = logging.getLogger('sanitycheck')
+# coloredlogs.install(level='INFO', logger=logger, fmt="%(levelname)s %(message)s")
```
why comment this out?

andrewboie commented on scripts/sanitycheck:
```diff
@@ -229,24 +228,19 @@ sys.path.insert(0, os.path.join(ZEPHYR_BASE, "scripts/"))
+logger.setLevel(logging.DEBUG)
+
+
+# log_format = "%(levelname)s %(message)s"
```
same

andrewboie commented on scripts/sanitycheck:
```diff
@@ -735,8 +706,8 @@ class DeviceHandler(Handler):
 
-
         while not self.device_is_available(self.instance.platform.name):
+            #logger.debug("Waiting for device to become available...")
```
keep or remove, but not comment out

nashif committed:
sanitycheck: remove commented out code

Remove commented out debug code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: fix reporting of timeouts

timeouts were not reported correctly and we were getting the default
'N/A' in case of a timeout.

Fixes #21438

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

chen-png commented on scripts/sanitycheck:
```diff
@@ -3448,9 +3463,12 @@ def log_info_file(instance):
+        log_info("{}".format(d_log))
     elif os.path.exists(h_log):
         log_info("{}".format(h_log))
     else:
```
should we judge whether it's the flash error or tests failed via judging whether the h_log is empty?

if it's the test case failed, I think we would more like to see hander.log,  if it's the flash error, then it should be device.log.
for now, it's all device.log.
